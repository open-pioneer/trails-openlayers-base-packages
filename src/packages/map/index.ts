// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0
export * from "./api";
