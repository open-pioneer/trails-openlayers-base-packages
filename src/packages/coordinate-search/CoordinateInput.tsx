// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import { computed, Reactive, reactive } from "@conterra/reactivity-core";
import {
    Box,
    Flex,
    InputGroup,
    InputRightAddon,
    Portal,
    Tooltip
} from "@open-pioneer/chakra-integration";
import { MapModelProps, useMapModel } from "@open-pioneer/map";
import { CommonComponentProps, useCommonComponentProps, useEvent } from "@open-pioneer/react-utils";
import { useReactiveSnapshot } from "@open-pioneer/reactivity";
import { NumberParserService, PackageIntl } from "@open-pioneer/runtime";
import { Coordinate } from "ol/coordinate";
import { get as getProjection, Projection, ProjectionLike, transform } from "ol/proj";
import { useIntl, useService } from "open-pioneer:react-hooks";
import { FC, useCallback, useEffect, useMemo, useRef, useState } from "react";
import { CoordinateInputField } from "./CoordinateInputField";
import { formatCoordinates, parseCoordinates, ParseResult } from "./coordinates";
import { ProjectionSelect } from "./ProjectionSelect";
import { usePlaceholder } from "./usePlaceholder";

const DEFAULT_PRECISION = 3;
const DEFAULT_PROJECTIONS = [
    {
        label: "WGS 84",
        value: getProjection("EPSG:4326")!,
        precision: 3
    },
    {
        label: "Web Mercator",
        value: getProjection("EPSG:3857")!,
        precision: 2
    }
];

/**
 * Dropdown item of projection selection with an optional coordinate precision
 */
export interface ProjectionInput {
    /**
     * Label to show the user.
     */
    label: string;

    /**
     * The map projection as projection or as string.
     */
    value: ProjectionLike;

    /**
     * The number of displayed decimal places.
     */
    precision?: number;
}

/**
 * Internal view of a (normalized) projection.
 * The projection has been looked up and optional values have been filled in.
 */
export interface ProjectionItem {
    /**
     * Label to show the user.
     */
    label: string;

    /**
     * The map projection.
     */
    value: Projection;

    /**
     * The number of displayed decimal places.
     */
    precision: number;
}

/**
 * Event type emitted when the user enters new coordinates or projection is changed by the user.
 */
export interface CoordinatesSelectEvent {
    /** coordinates in the projection of the object */
    coords: Coordinate;

    /** the projection of the coordinates. */
    projection: Projection;
}

/**
 * Props for the {@link CoordinateInput} component.
 */
export interface CoordinateInputProps extends CommonComponentProps, MapModelProps {
    /**
     * List of projection options, only projections that are known by the map as projection are shown.
     * Each projection can have an individual precision of coordinates.
     *
     * If no precision is given, the default precision is used.
     */
    projections?: ProjectionInput[];

    /**
     * Optional event that gets called if (valid) coordinates are entered or projection is changed by the user.
     */
    onSelect?: (event: CoordinatesSelectEvent) => void;

    /**
     * Optional event that gets called if the input is cleared.
     */
    onClear?: () => void;

    /**
     * Insert input value and overwrite user input.
     */
    input?: Coordinate;

    /**
     * Placeholder text to display when no input is present. Common usages:
     * * hint for the user ("enter coordinate here")
     * * example coordinate ("12.345 67.890")
     * * current mouse position
     *
     * If a coordinate is given, it has to be in the current projection of the map.
     */
    placeholder?: string | Coordinate;
}

/**
 * The `CoordinateInput` component can be used in an app to provide a validated input field for coordinates in a selected projection
 */
export const CoordinateInput: FC<CoordinateInputProps> = (props) => {
    const {
        onSelect: onSelectProp,
        onClear,
        projections = DEFAULT_PROJECTIONS,
        input,
        placeholder = ""
    } = props;
    const { containerProps } = useCommonComponentProps("coordinate-input", props);
    const { map } = useMapModel(props);
    const intl = useIntl();
    const mapProjection = useReactiveSnapshot(() => map?.projection, [map]);

    // Projection items (dropdown)
    const availableProjections = useProjectionItems(projections);
    const [selectedProjection, setSelectedProjection] = useState<ProjectionItem>(
        // choose first option initially
        availableProjections[0]!
    );

    // Input state
    const onSelect = useEvent((coordinatesResult: ParseResult) => {
        if (!onSelectProp || coordinatesResult.kind !== "success" || mapProjection == null) {
            return;
        }

        const coords = transform(
            coordinatesResult.coordinates,
            coordinatesResult.projection,
            mapProjection
        );
        onSelectProp({ coords: coords, projection: mapProjection });
    });
    const [coordinateSearchInput, setCoordinateSearchInput, validationResult] = useCoordinateState(
        input,
        mapProjection,
        selectedProjection,
        onSelect
    );
    const placeholderString = usePlaceholder(placeholder, mapProjection, selectedProjection);
    const isInputValid = validationResult.kind === "success" || validationResult.kind === "empty";
    const onEnter = useEvent(() => {
        onSelect(validationResult);
    });

    const portalElement = useRef<HTMLDivElement>(null);
    return (
        <Box {...containerProps}>
            <Portal>
                <div ref={portalElement} />
            </Portal>
            <Flex flexDirection={"row"} flexDir={"row"}>
                <Tooltip
                    label={
                        !isInputValid
                            ? intl.formatMessage({ id: validationResult.kind })
                            : undefined
                    }
                    hasArrow
                    placement="auto"
                    isOpen={!isInputValid}
                    className="coordinate-input-tooltip"
                >
                    <InputGroup className="coordinate-input-group">
                        <CoordinateInputField
                            coordinateSearchInput={coordinateSearchInput}
                            setCoordinateSearchInput={setCoordinateSearchInput}
                            placeholder={placeholder}
                            placeholderString={placeholderString}
                            onClear={onClear}
                            isInputValid={isInputValid}
                            onEnter={onEnter}
                        />
                        <InputRightAddon padding={"0px"} borderLeft={"0px"}>
                            <ProjectionSelect
                                portalElement={portalElement}
                                currentProjection={selectedProjection}
                                projections={availableProjections}
                                onProjectionChange={setSelectedProjection}
                            />
                        </InputRightAddon>
                    </InputGroup>
                </Tooltip>
            </Flex>
        </Box>
    );
};

/**
 * Returns the current text input and a callback to change it (used for interactive user input).
 * The current text may also change if the input prop changes (controlled usage).
 */
function useCoordinateState(
    inputProp: Coordinate | undefined,
    mapProjection: Projection | undefined,
    selectedProjection: ProjectionItem,
    onSelect: (validationResult: ParseResult) => void
): [string, (value: string) => void, ParseResult] {
    const intl = useIntl();
    const numberParser = useService<NumberParserService>("runtime.NumberParserService");

    const [model] = useState(() => new StateModel(intl, selectedProjection, numberParser));
    useEffect(() => {
        const triggerSelect =
            inputProp !== model.inputProp || selectedProjection !== model.selectedProjection;

        model.setI18n(intl, numberParser);
        model.setInputProp(inputProp);
        model.setSelectedProjection(selectedProjection);
        model.setMapProjection(mapProjection);

        if (triggerSelect) {
            const validationResult = model.validationResult;
            onSelect(validationResult);
        }
    }, [model, intl, numberParser, inputProp, selectedProjection, mapProjection, onSelect]);

    const { rawInput, validationResult } = useReactiveSnapshot(() => {
        return {
            rawInput: model.rawInput,
            validationResult: model.validationResult
        };
    }, [model]);
    const setInputText = useCallback(
        (inputText: string) => {
            model.setText(inputText);
        },
        [model]
    );
    return [rawInput, setInputText, validationResult];
}

class StateModel {
    #intl: Reactive<PackageIntl>;
    #selectedProjection: Reactive<ProjectionItem>;
    #mapProjection = reactive<Projection | undefined>();
    #inputProp = reactive<Coordinate | undefined>();
    #numberParser: Reactive<NumberParserService>;

    #rawInput = reactive("");
    #validationResult = computed(() => {
        return parseCoordinates(
            this.#rawInput.value,
            this.#numberParser.value,
            this.#selectedProjection.value.value
        );
    });

    constructor(
        intl: PackageIntl,
        selectedProjection: ProjectionItem,
        numberParser: NumberParserService
    ) {
        this.#intl = reactive(intl);
        this.#selectedProjection = reactive(selectedProjection);
        this.#numberParser = reactive(numberParser);
    }

    get inputProp() {
        return this.#inputProp.value;
    }

    get rawInput() {
        return this.#rawInput.value;
    }

    get validationResult() {
        return this.#validationResult.value;
    }

    get selectedProjection() {
        return this.#selectedProjection.value;
    }

    setI18n(intl: PackageIntl, numberParser: NumberParserService) {
        this.#intl.value = intl;
        this.#numberParser.value = numberParser;
    }

    setText(text: string) {
        this.#rawInput.value = text;
    }

    setSelectedProjection(value: ProjectionItem) {
        if (value !== this.#selectedProjection.value) {
            this.#selectedProjection.value = value;
            this.#updateInput();
        }
    }

    setInputProp(value: Coordinate | undefined) {
        if (value !== this.#inputProp.value) {
            this.#inputProp.value = value;
            this.#updateInput();
        }
    }

    setMapProjection(value: Projection | undefined) {
        if (value !== this.#mapProjection.value) {
            this.#mapProjection.value = value;
            this.#updateInput();
        }
    }

    #updateInput() {
        const inputProp = this.#inputProp.value;
        const mapProjection = this.#mapProjection.value;
        const selectedProjection = this.#selectedProjection.value;
        const intl = this.#intl.value;
        if (mapProjection && inputProp) {
            // Update state based on input prop.
            const transformed = transform(inputProp, mapProjection, selectedProjection.value);
            const formatted = formatCoordinates(transformed, selectedProjection.precision, intl);
            this.#rawInput.value = formatted;
        }
    }
}

/**
 * Builds the list of available projection items based on the provided list of projections
 */
function useProjectionItems(projections: ProjectionInput[]) {
    return useMemo(() => {
        // filter out projections that are not known
        const availableProjections: ProjectionItem[] = projections.flatMap((ele) => {
            if (getProjection(ele.value) != null)
                return {
                    label: ele.label,
                    value: getProjection(ele.value)!,
                    precision: ele.precision ?? DEFAULT_PRECISION
                };
            return [];
        });
        return availableProjections;
    }, [projections]);
}
