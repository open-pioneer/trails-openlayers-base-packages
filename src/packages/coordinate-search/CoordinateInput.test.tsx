// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import { nextTick } from "@conterra/reactivity-core";
import { MapContainer } from "@open-pioneer/map";
import { createServiceOptions, setupMap, waitForMapMount } from "@open-pioneer/map-test-utils";
import { PackageContextProvider } from "@open-pioneer/test-utils/react";
import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { Coordinate } from "ol/coordinate";
import { Projection } from "ol/proj";
import { disableReactActWarnings } from "test-utils";
import { beforeEach, expect, it, vi } from "vitest";
import { CoordinateInput } from "./CoordinateInput";
import {
    getClearButton,
    getCopyButton,
    getCurrentOptions,
    getCurrentOptionValues,
    showDropdown
} from "./test-utils";
import { NumberParserService } from "@open-pioneer/runtime";
import { NumberParser } from "@open-pioneer/core";

beforeEach(() => {
    disableReactActWarnings();
});

it("should successfully create a coordinate input component", async () => {
    const { mapId, injectedServices } = await setUp();
    render(
        <PackageContextProvider services={injectedServices}>
            <CoordinateInput mapId={mapId} data-testid="coordinate-input" />
        </PackageContextProvider>
    );

    // coordinate input is mounted
    const { coordsInputDiv } = await waitForCoordinateInput();
    expect(coordsInputDiv).toMatchSnapshot();

    // check coordinate input box is available
    expect(coordsInputDiv.tagName).toBe("DIV");
});

it("should successfully create a coordinate input component with additional css classes", async () => {
    const { mapId, injectedServices } = await setUp();

    render(
        <PackageContextProvider services={injectedServices}>
            <CoordinateInput mapId={mapId} className="test" data-testid="coordinate-input" />
        </PackageContextProvider>
    );

    const { coordsInputDiv } = await waitForCoordinateInput();
    expect(coordsInputDiv.classList.contains("test")).toBe(true);
    expect(coordsInputDiv.classList.contains("foo")).toBe(false);
});

it("should successfully create a coordinate input component with external input", async () => {
    const input = [761166, 6692084];
    const { mapId, injectedServices } = await setUp();

    render(
        <PackageContextProvider services={injectedServices} locale="de">
            <MapContainer mapId={mapId} data-testid="map" />
            <CoordinateInput mapId={mapId} data-testid="coordinate-input" input={input} />
        </PackageContextProvider>
    );

    await waitForMapMount("map");
    const { coordInput } = await waitForCoordinateInput();
    expect(coordInput.getAttribute("value")).toMatchInlineSnapshot('"6,838 51,398"');
});

it("should successfully create a coordinate input component with string placeholder", async () => {
    const { mapId, injectedServices } = await setUp();

    render(
        <PackageContextProvider services={injectedServices} locale="de">
            <MapContainer mapId={mapId} data-testid="map" />
            <CoordinateInput
                mapId={mapId}
                data-testid="coordinate-input"
                placeholder={"test placeholder"}
            />
        </PackageContextProvider>
    );

    await waitForMapMount("map");
    const { coordInput } = await waitForCoordinateInput();
    expect(coordInput.getAttribute("placeholder")).toMatchInlineSnapshot('"test placeholder"');
});

it("should successfully create a coordinate input component without a string placeholder", async () => {
    const { mapId, injectedServices } = await setUp();

    render(
        <PackageContextProvider services={injectedServices} locale="de">
            <MapContainer mapId={mapId} data-testid="map" />
            <CoordinateInput mapId={mapId} data-testid="coordinate-input" />
        </PackageContextProvider>
    );

    await waitForMapMount("map");
    const { coordInput } = await waitForCoordinateInput();
    expect(coordInput.getAttribute("placeholder")).toMatchInlineSnapshot('""');
});

it("should successfully create a coordinate input component with coordinate placeholder", async () => {
    const coord: Coordinate = [408000, 5600000];

    const { mapId, injectedServices } = await setUp();
    render(
        <PackageContextProvider services={injectedServices} locale="de">
            <MapContainer mapId={mapId} data-testid="map" />
            <CoordinateInput mapId={mapId} data-testid="coordinate-input" placeholder={coord} />
        </PackageContextProvider>
    );

    await waitForMapMount("map");
    const { coordInput } = await waitForCoordinateInput();
    expect(coordInput.getAttribute("placeholder")).toMatchInlineSnapshot('"3,665 44,863"');
});

it("should successfully create a coordinate input component with known projections", async () => {
    const { mapId, injectedServices } = await setUp();
    const projections = [
        {
            label: "EPSG:25832",
            value: "EPSG:25832"
        },
        {
            label: "WGS 84",
            value: "EPSG:4326"
        },
        {
            label: "Web Mercator",
            value: "EPSG:3857"
        },
        {
            label: "Unknown to ignore",
            value: "EPSG:99999"
        }
    ];

    render(
        <PackageContextProvider services={injectedServices}>
            <MapContainer mapId={mapId} data-testid="map" />
            <CoordinateInput
                mapId={mapId}
                data-testid="coordinate-input"
                projections={projections}
            />
        </PackageContextProvider>
    );

    await waitForMapMount("map");
    const { projSelect } = await waitForCoordinateInput();
    showDropdown(projSelect);
    const options = getCurrentOptions();
    const values = getCurrentOptionValues(options);

    expect(values).toStrictEqual(["EPSG:25832", "WGS 84", "Web Mercator"]);
});

it("should successfully create a coordinate input component with default projections", async () => {
    const { mapId, injectedServices } = await setUp();

    render(
        <PackageContextProvider services={injectedServices}>
            <MapContainer mapId={mapId} data-testid="map" />
            <CoordinateInput mapId={mapId} data-testid="coordinate-input" />
        </PackageContextProvider>
    );

    await waitForMapMount("map");
    const { projSelect } = await waitForCoordinateInput();
    showDropdown(projSelect);
    const options = getCurrentOptions();
    const values = getCurrentOptionValues(options);

    expect(values).toStrictEqual(["WGS 84", "Web Mercator"]);
});

it("should update coordinates in selected option", async () => {
    const { mapId, injectedServices } = await setUp();
    const initialInput = [851594.11, 6789283.95];
    const updatedInput = [860000, 6900000.95];

    function componentContent(input: number[]) {
        return (
            <PackageContextProvider services={injectedServices}>
                <MapContainer mapId={mapId} data-testid="map" />
                <CoordinateInput mapId={mapId} data-testid="coordinate-input" input={input} />
            </PackageContextProvider>
        );
    }

    const { rerender } = render(componentContent(initialInput));

    await waitForMapMount("map");
    const { coordInput } = await waitForCoordinateInput();
    expect(coordInput.getAttribute("value")).toMatchInlineSnapshot('"7.650 51.940"');

    rerender(componentContent(updatedInput));
    await nextTick(); // slightly delayed by useReactiveSnapshot()
    expect(coordInput.getAttribute("value")).toMatchInlineSnapshot(`"7.726 52.549"`);
});

it("should display transformed coordinates in selected option", async () => {
    const user = userEvent.setup();
    const input = [851594.11, 6789283.95];
    const { mapId, injectedServices } = await setUp();

    render(
        <PackageContextProvider services={injectedServices}>
            <MapContainer mapId={mapId} data-testid="map" />
            <CoordinateInput mapId={mapId} data-testid="coordinate-input" input={input} />
        </PackageContextProvider>
    );

    await waitForMapMount("map");
    const { coordInput, projSelect } = await waitForCoordinateInput();
    showDropdown(projSelect);

    let options = getCurrentOptions();
    const option4326 = options.find((option) => option.textContent === "WGS 84");
    if (!option4326) {
        throw new Error("EPSG 4326 missing in options");
    }
    await user.click(option4326);

    expect(coordInput.getAttribute("value")).toMatchInlineSnapshot('"7.650 51.940"'); //should display EPSG 4326

    showDropdown(projSelect);
    options = getCurrentOptions();
    const option3857 = options.find((option) => option.textContent === "Web Mercator");
    if (!option3857) {
        throw new Error("EPSG 3857 missing in options");
    }
    await user.click(option3857);
    expect(coordInput.getAttribute("value")).toMatchInlineSnapshot('"851,594.11 6,789,283.95"'); //should display EPSG 3857
});

it(
    "should successfully call onSelect after entering a new input",
    {
        timeout: 10000
    },
    async () => {
        const user = userEvent.setup();
        const { mapId, injectedServices } = await setUp();

        let searchedCoords: Coordinate = [];
        let callbackProj: Projection | undefined = undefined;
        render(
            <PackageContextProvider services={injectedServices}>
                <MapContainer mapId={mapId} data-testid="map" />
                <CoordinateInput
                    mapId={mapId}
                    data-testid="coordinate-input"
                    onSelect={({ coords, projection }) => {
                        searchedCoords = coords;
                        callbackProj = projection;
                    }}
                    onClear={() => {}}
                />
            </PackageContextProvider>
        );

        await waitForMapMount("map");
        const { coordInput } = await waitForCoordinateInput();
        await user.type(coordInput, "7 51{enter}");
        expect(searchedCoords).toStrictEqual([779236.4355529151, 6621293.722740165]);
        expect(callbackProj).toBeDefined();
        expect(callbackProj!.getCode()).toBe("EPSG:3857");
    }
);

it(
    "should not call onSelect after entering an invalid input",
    {
        timeout: 10000
    },
    async () => {
        const user = userEvent.setup();
        const { mapId, injectedServices } = await setUp();

        let callbackProj: Projection | undefined = undefined;
        let searchedCoords: Coordinate = [];
        let called = false;
        render(
            <PackageContextProvider services={injectedServices}>
                <MapContainer mapId={mapId} data-testid="map" />
                <CoordinateInput
                    mapId={mapId}
                    data-testid="coordinate-input"
                    onSelect={({ coords, projection }) => {
                        searchedCoords = coords;
                        callbackProj = projection;
                        called = true;
                    }}
                    onClear={() => {}}
                />
            </PackageContextProvider>
        );

        await waitForMapMount("map");
        const { coordInput } = await waitForCoordinateInput();
        await user.type(coordInput, "a b{enter}");
        expect(searchedCoords).toStrictEqual([]);
        expect(callbackProj).toBeUndefined();
        expect(called).toBeFalsy();
    }
);

it(
    "should successfully call onClear if clear button is clicked",
    {
        timeout: 10000
    },
    async () => {
        const user = userEvent.setup();
        const { mapId, injectedServices } = await setUp();

        let cleared: boolean = false;
        render(
            <PackageContextProvider services={injectedServices}>
                <MapContainer mapId={mapId} data-testid="map" />
                <CoordinateInput
                    mapId={mapId}
                    data-testid="coordinate-input"
                    onSelect={() => {}}
                    onClear={() => {
                        cleared = true;
                    }}
                />
            </PackageContextProvider>
        );

        await waitForMapMount("map");
        const { coordInput, coordinateInputGroup } = await waitForCoordinateInput();

        await user.type(coordInput, "4 5{enter}");
        const clearButton = getClearButton(coordinateInputGroup);
        expect(coordInput.getAttribute("value")).toMatchInlineSnapshot('"4 5"');
        expect(cleared).toBe(false);

        await user.click(clearButton);
        expect(coordInput.getAttribute("value")).toMatchInlineSnapshot('""');
        expect(cleared).toBe(true);
    }
);

it("should successfully copy to clipboard if copy button is clicked", async () => {
    const user = userEvent.setup();
    const { mapId, injectedServices } = await setUp();
    let copiedText = "";

    render(
        <PackageContextProvider services={injectedServices}>
            <MapContainer mapId={mapId} data-testid="map" />
            <CoordinateInput
                mapId={mapId}
                data-testid="coordinate-input"
                placeholder={[405000, 58000000]}
            />
        </PackageContextProvider>
    );

    await waitForMapMount("map");
    const { coordinateInputGroup } = await waitForCoordinateInput();

    vi.spyOn(navigator, "clipboard", "get").mockReturnValue({
        writeText(input: string) {
            copiedText = input;
            return new Promise<void>((resolve) => resolve());
        },
        readText(): Promise<string> {
            return new Promise<string>((resolve) => resolve(""));
        },
        write(): Promise<void> {
            return new Promise<void>((resolve) => resolve());
        },
        addEventListener(): void {},
        dispatchEvent(): boolean {
            return false;
        },
        removeEventListener(): void {},
        read(): Promise<ClipboardItems> {
            return new Promise<ClipboardItems>((resolve) => resolve([]));
        }
    });
    const copyButton = getCopyButton(coordinateInputGroup);
    await user.click(copyButton);
    expect(copiedText).toBe("3.638 89.987");
});

it("should validate the input and show the correct tooltip message", async () => {
    const user = userEvent.setup();
    const tooltipHelper = createTooltipHelper();
    const { mapId, injectedServices } = await setUp("de");

    render(
        <PackageContextProvider services={injectedServices} locale="de">
            <MapContainer mapId={mapId} data-testid="map" />
            <CoordinateInput
                mapId={mapId}
                data-testid="coordinate-input"
                onSelect={() => {}}
                onClear={() => {}}
            />
        </PackageContextProvider>
    );

    await waitForMapMount("map");
    const { coordInput } = await waitForCoordinateInput();

    await user.clear(coordInput);
    await user.click(coordInput);
    await user.paste("ab");
    const tooltip1 = await tooltipHelper.waitForChange();
    expect(tooltip1).toBe("tooltip.space");

    await user.clear(coordInput);
    await user.click(coordInput);
    await user.paste("a  a");
    const tooltip2 = await tooltipHelper.waitForChange();
    expect(tooltip2).toBe("tooltip.spaceOne");

    await user.clear(coordInput);
    await user.click(coordInput);
    await user.paste("a ");
    const tooltip3 = await tooltipHelper.waitForChange();
    expect(tooltip3).toBe("tooltip.2coords");

    await user.clear(coordInput);
    await user.click(coordInput);
    await user.paste("a,b c,d");
    const tooltip4 = await tooltipHelper.waitForChange();
    expect(tooltip4).toBe("tooltip.invalidNumbers");

    await user.clear(coordInput);
    await user.click(coordInput);
    await user.paste("a b c");
    const tooltip5 = await tooltipHelper.waitForChange();
    expect(tooltip5).toBe("tooltip.spaceOne");

    await user.clear(coordInput);
    await user.click(coordInput);
    await user.paste("a b");
    const tooltip6 = await tooltipHelper.waitForChange();
    expect(tooltip6).toBe("tooltip.invalidNumbers");
});

it("should validate if the user input is within the extent of the selected projection", async () => {
    const user = userEvent.setup();
    const tooltipHelper = createTooltipHelper();
    const { mapId, injectedServices } = await setUp();

    render(
        <PackageContextProvider services={injectedServices} locale="de">
            <MapContainer mapId={mapId} data-testid="map" />
            <CoordinateInput
                mapId={mapId}
                data-testid="coordinate-input"
                onSelect={() => {}}
                onClear={() => {}}
            />
        </PackageContextProvider>
    );

    await waitForMapMount("map");
    const { coordInput } = await waitForCoordinateInput();

    await user.clear(coordInput);
    await user.click(coordInput);
    await user.paste("200 100");
    const tooltip8 = await tooltipHelper.waitForChange();
    expect(tooltip8).toBe("tooltip.extent");
});

it("should validate input property is within the extent of the selected projection", async () => {
    const user = userEvent.setup();
    const { mapId, injectedServices } = await setUp();

    render(
        <PackageContextProvider services={injectedServices} locale="de">
            <MapContainer mapId={mapId} data-testid="map" />
            <CoordinateInput
                mapId={mapId}
                data-testid="coordinate-input"
                onSelect={() => {}}
                onClear={() => {}}
                input={[3495814159, 56963568019]}
            />
        </PackageContextProvider>
    );

    await waitForMapMount("map");
    const { coordInput } = await waitForCoordinateInput();
    await user.click(coordInput);
    const tooltip = getCurrentTooltipText();
    expect(tooltip).toBe("tooltip.extent");
});

it("should not show copy button on string placeholder", async () => {
    const { mapId, injectedServices } = await setUp();

    render(
        <PackageContextProvider services={injectedServices}>
            <MapContainer mapId={mapId} data-testid="map" />
            <CoordinateInput mapId={mapId} data-testid="coordinate-input" placeholder={"test"} />
        </PackageContextProvider>
    );

    await waitForMapMount("map");
    const { coordinateInputGroup } = await waitForCoordinateInput();
    expect(() => getCopyButton(coordinateInputGroup)).toThrowError(/^buttons not rendered$/);
    expect(() => getClearButton(coordinateInputGroup)).toThrowError(/^buttons not rendered$/);
});

it("should show copy button on coordinate placeholder", async () => {
    const { mapId, injectedServices } = await setUp();

    render(
        <PackageContextProvider services={injectedServices}>
            <MapContainer mapId={mapId} data-testid="map" />
            <CoordinateInput
                mapId={mapId}
                data-testid="coordinate-input"
                placeholder={[405000, 58000000]}
            />
        </PackageContextProvider>
    );

    await waitForMapMount("map");
    const { coordinateInputGroup } = await waitForCoordinateInput();
    const copyButton = getCopyButton(coordinateInputGroup);
    expect(copyButton).toBeDefined();
    expect(() => getClearButton(coordinateInputGroup)).toThrowError(/^clear button not rendered$/);
});

it("should show clear button on coordinate placeholder", async () => {
    const user = userEvent.setup();
    const { mapId, injectedServices } = await setUp();

    render(
        <PackageContextProvider services={injectedServices}>
            <MapContainer mapId={mapId} data-testid="map" />
            <CoordinateInput
                mapId={mapId}
                data-testid="coordinate-input"
                onSelect={() => {}}
                onClear={() => {}}
            />
        </PackageContextProvider>
    );

    await waitForMapMount("map");
    const { coordInput, coordinateInputGroup } = await waitForCoordinateInput();

    await user.type(coordInput, "4 5{enter}");

    const clearButton = getClearButton(coordinateInputGroup);
    expect(clearButton).toBeDefined();
    expect(() => getCopyButton(coordinateInputGroup)).toThrowError(/^copy button not rendered$/);
});

async function waitForCoordinateInput() {
    const { coordsInputDiv, coordInput, coordinateInputGroup, projSelect } = await waitFor(
        async () => {
            const coordsInputDiv = await screen.findByTestId("coordinate-input");

            const coordinateInputGroup = coordsInputDiv.querySelector(".coordinate-input-group");
            if (!coordinateInputGroup) {
                throw new Error("coordinate input group not rendered");
            }

            const coordInputDiv = coordinateInputGroup.querySelector(
                ".coordinate-input-field-group"
            );
            if (!coordInputDiv) {
                throw new Error("coordinate input field  group not rendered");
            }

            const coordInput = coordInputDiv.querySelector(".chakra-input");
            if (!coordInput) {
                throw new Error("coordinate input field not rendered");
            }

            const projSelect: HTMLElement | null = coordinateInputGroup.querySelector(
                ".coordinate-input-select--has-value"
            );
            if (!projSelect) {
                throw new Error("coordinate input projection select not rendered");
            }

            return { coordsInputDiv, coordInput, coordinateInputGroup, projSelect };
        }
    );

    return { coordsInputDiv, coordInput, coordinateInputGroup, projSelect };
}

// Returns undefined if no unique tooltip was found
function getCurrentTooltipText(): string | undefined {
    const tooltips = document.getElementsByClassName("coordinate-input-tooltip");
    if (!tooltips.length || tooltips.length > 1) {
        return undefined;
    }
    return (tooltips[0] as HTMLElement).textContent ?? undefined;
}

function createTooltipHelper() {
    let currentTooltip = getCurrentTooltipText();
    return {
        /** Waits until the tooltip has a different content (including undefined) from the last time it was checked. */
        async waitForChange() {
            const newTooltip = await waitFor(() => {
                const newTooltip = getCurrentTooltipText();
                if (newTooltip === currentTooltip) {
                    throw new Error(`Tooltip has not changed, still '${currentTooltip}'`);
                }
                return newTooltip;
            });
            currentTooltip = newTooltip;
            return newTooltip;
        },
        currentTooltip() {
            return currentTooltip;
        }
    };
}

async function setUp(locale: string = "en") {
    const { mapId, registry } = await setupMap();
    const numberParser = new NumberParser(locale);
    const numberParserService = {
        parseNumber: (number) => {
            return numberParser.parse(number);
        }
    } satisfies Partial<NumberParserService>;

    const injectedServices = {
        ...createServiceOptions({ registry }),
        "runtime.NumberParserService": numberParserService
    };

    return { mapId, injectedServices };
}
