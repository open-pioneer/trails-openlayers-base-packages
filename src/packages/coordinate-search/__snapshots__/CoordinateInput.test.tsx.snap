// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should successfully create a coordinate input component 1`] = `
<div
  class="coordinate-input css-0"
  data-testid="coordinate-input"
  data-theme="light"
>
  <div
    class="css-1xhj18k"
    data-theme="light"
  >
    <div
      class="chakra-input__group coordinate-input-group css-1y0e7gb"
      data-group="true"
      data-theme="light"
    >
      <div
        class="chakra-input__group coordinate-input-field-group css-1y0e7gb"
        data-group="true"
        data-theme="light"
      >
        <input
          aria-label="coordinateInput.ariaLabel"
          class="chakra-input css-1r29ykt"
          data-theme="light"
          placeholder=""
          type="text"
          value=""
        />
      </div>
      <div
        class="chakra-input__right-addon css-vqkjic"
        data-theme="light"
      >
        <div
          class="coordinate-input-select--has-value css-79elbk"
          data-theme="light"
        >
          <span
            class="css-1f43avz-a11yText-A11yText"
            id="react-select-2-live-region"
          />
          <span
            aria-atomic="false"
            aria-live="polite"
            aria-relevant="additions text"
            class="css-1f43avz-a11yText-A11yText"
            role="log"
          />
          <div
            class="coordinate-input-select__control css-u1k0x6"
            data-theme="light"
          >
            <div
              class="coordinate-input-select__value-container coordinate-input-select__value-container--has-value css-odo7kh"
              data-theme="light"
            >
              <div
                class="coordinate-input-select__single-value css-1xa1gs2"
                data-theme="light"
              >
                WGS 84
              </div>
              <input
                aria-autocomplete="list"
                aria-expanded="false"
                aria-haspopup="true"
                aria-label="coordinateInput.ariaLabel"
                aria-readonly="true"
                class="css-mohuvp-dummyInput-DummyInput"
                id="react-select-2-input"
                inputmode="none"
                role="combobox"
                tabindex="0"
                value=""
              />
            </div>
            <div
              class="coordinate-input-select__indicators css-hfbj6y"
              data-theme="light"
            >
              <hr
                aria-orientation="vertical"
                class="chakra-divider coordinate-input-select__indicator-separator css-ejwint"
                data-theme="light"
              />
              <div
                aria-hidden="true"
                class="coordinate-input-select__indicator coordinate-input-select__dropdown-indicator css-4v68kg"
                data-theme="light"
              >
                <svg
                  aria-hidden="true"
                  class="chakra-icon css-onkibi"
                  data-theme="light"
                  focusable="false"
                  role="presentation"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                    fill="currentColor"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
